### ISSUE TITLE
Difficult signalling behavior for URLLoader responses

### ISSUE THREAD

**Issue by johngro** [domokit/mojo#497]
Repro

1) Create a request for some piece of content using a URLLoader
2) Grab the handle to the response body from the URLResponse when the callback executes (it is a ScopedDataPipeConsumerHandle)
3) Attempt to read some data (say, 16KB) from the body using ReadDataRaw and passing the MOJO_READ_DATA_FLAG_ALL_OR_NONE flag.

Expected:
If the data is not yet in the cache, the ReadDataRaw call should return MOJO_RESULT_SHOULD_WAIT
Actual:
If the data is not yet in the cache, the ReadDataRaw call will return MOJO_RESULT_OUT_OF_RANGE

4) Now call wait on the handle, passing MOJO_HANDLE_SIGNAL_READABLE

Expected, and actual, but unfortunate and difficult:
The wait call immediately unblocks.  The handle is readable, because it has some bytes available for read, it just does not have 16KB (or whatever).  Calling ReadDataRaw will continue to return MOJO_RESULT_OUT_OF_RANGE until it can satisfy the ALL_OR_NONE condition of the read.

5) Finally, keep going, and eventually attempt to read past the end of the body.
Expected:
An error code indicating that the data cannot be read (not now, not ever) because it is past the end of the file.  Perhaps MOJO_RESULT_OUT_OF_RANGE, or MOJO_RESULT_UNAVAILABLE.
Actual:
The read operation returns MOJO_RESULT_PRECONDITION_FAILED

Issues;
The first error code is just confusing.  The read is not out of range, it just cannot be satisfied yet because the data is not in the cache.  I would expect SHOULD_WAIT, not OUT_OF_RANGE.

The problem with returning SHOULD_WAIT, however, is that waiting will do you no good.  Attempting to wait will simply spin because the handle is signaled for read any time you have even a single byte to read.  Having a mechanism to indicate that you want to wait until there are X bytes to read, or the stream has ended (whichever comes first), would be very helpful.  Being able to query the handle to know how many bytes may be immediately read without blocking, and whether or not there may be more bytes coming after that, would also be helpful.

Finally, the PRECONDITION_FAILED error code just feels strange.  I could understand this error if I did something like pass NULL for the pointer to read into, but if I'm attempting to read pas the end of the stream it seems like there should be a different error.

**Comment by johngro** [comment]
@j9brown @chrismckillop @viettrungluu-cr @dalesat
